src/work/Work.cpp

```diff
@@ -4,434 +4,224 @@
-        return fmt::format("Awaiting {:d}/{:d} prerequisites of: {:s}",
-                           total - i, total, getUniqueName());
-    }
-    case WORK_RUNNING:
-        return fmt::format("Running: {:s}", getUniqueName());
-    case WORK_SUCCESS:
-        return fmt::format("Succeded: {:s}", getUniqueName());
-    case WORK_FAILURE_RETRY:
-    {
-        auto eta = getRetryETA();
-        return fmt::format("Retrying in {:d} sec: {:s}", eta, getUniqueName());
-    }
-    case WORK_FAILURE_RAISE:
-    case WORK_FAILURE_FATAL:
-        return fmt::format("Failed: {:s}", getUniqueName());
-    default:
-        assert(false);
-        return "";
-    }
-}
-
-uint64_t
-Work::getRetryETA() const
-{
-    uint64_t now = mApp.timeNow();
-    uint64_t retry =
-        mRetryTimer ? VirtualClock::to_time_t(mRetryTimer->expiry_time()) : 0;
-    return now > retry ? 0 : retry - now;
-}
-
-VirtualClock::duration
-Work::getRetryDelay() const
-{
-    // Cap to 4096sec == a little over an hour.
-    uint64_t m = 2 << std::min(uint64_t(12), uint64_t(mRetries));
-    return std::chrono::seconds(rand_uniform<uint64_t>(1ULL, m));
-}
-
-size_t
-Work::getMaxRetries() const
-{
-    return mMaxRetries;
-}
-
-std::string
-Work::stateName(State st)
-{
-    switch (st)
-    {
-    case WORK_PENDING:
-        return "WORK_PENDING";
-    case WORK_RUNNING:
-        return "WORK_RUNNING";
-    case WORK_SUCCESS:
-        return "WORK_SUCCESS";
-    case WORK_FAILURE_RETRY:
-        return "WORK_FAILURE_RETRY";
-    case WORK_FAILURE_RAISE:
-        return "WORK_FAILURE_RAISE";
-    case WORK_FAILURE_FATAL:
-        return "WORK_FAILURE_FATAL";
-    default:
-        throw std::runtime_error("Unknown Work::State");
+        return fmt::format("Work {:s} has incomplete children: {:d}/{:d}",
+                           getName(), incomplete, mChildren.size());
     }
-}
-
-std::function<void(asio::error_code const& ec)>
-Work::callComplete()
-{
-    std::weak_ptr<Work> weak(
-        std::static_pointer_cast<Work>(shared_from_this()));
-    return [weak](asio::error_code const& ec) {
-        auto self = weak.lock();
-        if (!self)
-        {
-            return;
-        }
-        self->complete(ec ? WORK_COMPLETE_FAILURE : WORK_COMPLETE_OK);
-    };
-}
-
-void
-Work::scheduleRun()
-{
-    if (mScheduled)
-    {
-        return;
-    }
-
-    std::weak_ptr<Work> weak(
-        std::static_pointer_cast<Work>(shared_from_this()));
-    CLOG(DEBUG, "Work") << "scheduling run of " << getUniqueName();
-    mScheduled = true;
-    mApp.getClock().getIOService().post([weak]() {
-        auto self = weak.lock();
-        if (!self)
-        {
-            return;
-        }
-        self->mScheduled = false;
-        self->run();
-    });
+    return BasicWork::getStatus();
 }
 
 void
-Work::scheduleComplete(CompleteResult result)
+Work::shutdown()
 {
-    if (mScheduled)
+    for (auto const& c : mChildren)
     {
-        return;
+        c->shutdown();
     }
```

Comment:
* Children are removed, because `BasicWork::shutdown` calls `reset`, which clears children. 
* Done (though it required exposing "destructing" state publicly, but I think it's a vital piece of information for the implementers anyway)